Implement message dispatch system

cwillisf opened the pull request:
### Proposed Changes

This message dispatch system allows a "service" to register itself under a particular name, and then anyone can call any method on that service from any context (the main "window" thread or any worker). Return values are passed back through promise resolution.

### Reason for Changes

This will make it much easier to put extensions inside workers: we can expose any services needed by extensions through this message dispatch system, but otherwise extensions will be unable to interact with the editor. If we find that it's necessary, we could even add some form of permissions to the message dispatch system with relatively little work.

### Test Coverage

New tests cover the basics of the message dispatch system. Because the system is built around web workers, which aren't present in our Node.js-based testing process, I added the `tiny-worker` library. This library emulates web workers by running multiple Node processes.

Unfortunately this exposes the difference between webpack's build-time handling and Node's run-time handling of `require` statements. The `dispatch-test-worker-shim.js` file monkey-patches `require` **in the worker only** to deal with these differences.

Commit:
Implement message dispatch system

This message dispatch system allows a "service" to register itself under
a particular name, and then anyone can call any method on that service
from any context (the main "window" thread or any worker). Return values
are passed back through promise resolution.

rschamp commented on src/dispatch/central-dispatch.js:
```diff
@@ -0,0 +1,173 @@
+const dispatch = new CentralDispatch();
+module.exports = dispatch;
+self.Scratch = self.Scratch || {};
+self.Scratch.dispatch = dispatch;
```
Is this left in for debugging? Seems like it should be something the app does.

Commit:
Add tests for message dispatch system; fix bugs

The tests run using TinyWorker, which emulates web workers on Node.
There are quite a few quirks in that situation due to the differences
between Node and Webpack as well as the differences between TinyWorker
and real Web Workers.

The tests also exposed a few bugs in the dispatch system, which have now
been fixed. Most notably, if a method called through the dispatch system
throws an exception that exception will now be passed back to the
caller. Previously the exception would escape the dispatch system and
the caller would never hear any response at all.

Commit:
Fix exception message

Commit:
Allow 'then' on service registration

This allows a service to postpone communication with other services
until it can be sure that it's registered with central dispatch. Service
registration on the main thread always happens immediately, but that
version of `setService` still returns a Promise for consistency.

cwillisf commented:
Rebased on `develop` @ e249c5c2525316a369a2f6a5220467eb6c1d60b5 to fix `package.json` conflicts

rschamp commented on src/dispatch/central-dispatch.js:
```diff
@@ -0,0 +1,188 @@
+     * @returns {Promise} - a promise which will resolve once the service is registered.
+     */
+    setService (service, provider) {
+        return new Promise(resolve => {
```
We should add an explanatory comment about why this returns a Promise (for consistency).

rschamp commented on src/dispatch/central-dispatch.js:
```diff
@@ -0,0 +1,188 @@
+     * @private
+     */
+    _onMessage (worker, event) {
+        const [service, method, callbackId, ...args] = /** @type {[string, string, *]} */ event.data;
```
Maybe `event.data` should come as an object with a `type` key. Either way we should `try...except` here in case the message isn't well formed.

rschamp commented on src/dispatch/central-dispatch.js:
```diff
@@ -0,0 +1,188 @@
+     * @param {*} [error] - if this is truthy, reject the callback promise with this value.
+     * @private
+     */
+    _callback (callbackId, result, error) {
```
`result, error` looks reversed to me.

rschamp reviewed (changes requested):
@cwillisf Looks great overall! Just a few notes from our walkthrough.

Commit:
Refactor common code in {central,worker}-dispatch

Central dispatch and worker dispatch share most of their code now by
inheriting from a new shared dispatch class.

Also, the format of passed messages has been altered to make it easier
to understand in a debugger.

rschamp reviewed (approved):
Looks good @cwillisf! Can't wait to see this in action.

rschamp commented:
@paulkaplan @fsih @ericrosenbaum should also take a look at this, at least just to be familiar with what's going in for the extensions work.
